File: hosts/intellij/src/main/kotlin/com/ohmycaptain/core/NodeExecutableLocator.kt
package com.ohmycaptain.core

import com.ohmycaptain.logging.loggerFor
import java.io.File

/**
 * 시스템에 설치된 Node.js 실행 파일 경로를 탐색하는 헬퍼.
 *
 * 관리되는 Node 버전(nvm/fnm) 우선, 없으면 시스템 PATH 로 fallback.
 * IntelliJ 가 띄워진 환경(맥OS GUI 런처)에서는 셸 환경변수가 일부 누락되므로
 * 직접 파일시스템을 훑는 방식으로 첫 번째 후보를 만든다.
 *
 * 탐색 우선순위:
 * 1. nvm default alias (`~/.nvm/alias/default`) → 사용자가 명시적으로 지정한 기본 버전
 * 2. nvm versions 디렉터리 스캔 → v24 우선, 그 외에는 버전명 내림차순
 * 3. fnm default alias 및 시스템 PATH (`/usr/local/bin`, `/opt/homebrew/bin`) — `which`/`where` 호출
 *
 * 어느 단계에서도 못 찾으면 null. [CoreApplicationService.startCore] 가 사용자에게 에러를 노출한다.
 */
internal object NodeExecutableLocator {

    private val log = loggerFor(NodeExecutableLocator::class)

    /**
     * Node 실행 파일 절대 경로 또는 null.
     *
     * 어떤 fallback 단계에서 찾았는지 디버그 로그로 남긴다 — 사용자 환경별 동작 차이를 진단하는 데 결정적.
     */
    fun find(): String? {
        findFromNvmDefault()?.let {
            log.debug { "[OMC] Node 발견: nvm default alias ($it)" }
            return it
        }
        findFromNvmVersions()?.let {
            log.debug { "[OMC] Node 발견: nvm versions 스캔 ($it)" }
            return it
        }
        findFromPathFallback()?.let {
            log.debug { "[OMC] Node 발견: PATH fallback ($it)" }
            return it
        }
        log.warn("[OMC] Node 실행 파일을 모든 fallback 단계에서 찾지 못함")
        return null
    }

    /**
     * 1) nvm default alias 파일에서 기본 버전을 읽고, 해당 버전의 node 바이너리 경로를 반환.
     *
     * `~/.nvm/alias/default` 는 `lts/iron`, `v24.0.0`, `node` 등 다양한 형태가 들어있을 수 있어
     * `versions/node/v{버전}` 디렉터리와 prefix 매칭하는 가장 최신 버전을 고른다.
     */
    private fun findFromNvmDefault(): String? {
        val nvmDir = nvmDir()
        val aliasFile = File("$nvmDir/alias/default")
        if (!aliasFile.exists()) return null
        val versionPrefix = aliasFile.readText().trim()

        val resolved = File("$nvmDir/versions/node")
            .listFiles()
            ?.filter {
                it.isDirectory &&
                    it.name.removePrefix("v").startsWith(versionPrefix) &&
                    File(it, "bin/node").exists()
            }
            ?.maxByOrNull { it.name }
            ?.let { "${it.absolutePath}/bin/node" }

        return resolved?.takeIf { File(it).exists() }
    }

    /**
     * 2) nvm 설치 디렉터리 전체 스캔.
     *
     * 정렬: v24 가 가장 우선(현재 권장 LTS), 그 외에는 버전 문자열 내림차순(v22 > v20 > v18 ...).
     * v24 우선 가중치는 `removePrefix("v").split(".").firstOrNull()?.toIntOrNull() == 24` 비교로 부여.
     */
    private fun findFromNvmVersions(): String? {
        val candidates = File("${nvmDir()}/versions/node")
            .listFiles()
            ?.filter { it.isDirectory && File(it, "bin/node").exists() }
            ?.sortedWith(
                compareByDescending<File> {
                    it.name.removePrefix("v").split(".").firstOrNull()?.toIntOrNull() == 24
                }.thenByDescending { it.name }
            )
            ?.map { "${it.absolutePath}/bin/node" }
            .orEmpty()
        return candidates.firstOrNull()
    }

    /**
     * 3) nvm 미사용 환경 fallback — `which node` (Windows: `where node`) 를 직접 호출.
     *
     * 환경변수 PATH 를 OS 기본 + fnm + Homebrew 경로로 보강한다 — IntelliJ GUI 런치 시
     * 사용자 셸의 PATH 가 그대로 상속되지 않기 때문.
     */
    private fun findFromPathFallback(): String? {
        val home = System.getProperty("user.home")
        val extraPaths = listOfNotNull(
            "$home/.local/share/fnm/aliases/default/bin",
            "/usr/local/bin",
            "/opt/homebrew/bin",
        )
        val pathEnv = (System.getenv("PATH") ?: "") + ":" + extraPaths.joinToString(":")
        val isWindows = System.getProperty("os.name").startsWith("Windows")
        val name = if (isWindows) "node.exe" else "node"

        return try {
            val pb = ProcessBuilder(if (isWindows) "where" else "which", name)
            pb.environment()["PATH"] = pathEnv
            pb.start().inputStream.bufferedReader().readLine()?.trim()?.takeIf { it.isNotEmpty() }
        } catch (e: Exception) {
            // which/where 자체가 없거나 실행 권한 문제 — 매우 드문 환경 이슈. 디버깅 가치 있음.
            log.debug(e) { "[OMC] PATH fallback 의 which/where 호출 실패" }
            null
        }
    }

    /** NVM_DIR 환경변수가 있으면 그걸, 없으면 ~/.nvm 사용. */
    private fun nvmDir(): String =
        System.getenv("NVM_DIR") ?: "${System.getProperty("user.home")}/.nvm"
}

File: hosts/vscode/src/core/NodeExecutableLocator.ts
// 시스템에 설치된 Node.js 실행 파일 경로 탐색기.
// IntelliJ 측 NodeExecutableLocator.kt 의 nvm/fnm/PATH fallback 로직을 그대로 포팅한다.
//
// 마지막 fallback 으로 process.execPath (extension host 의 Electron Node) 도 시도해
// 사용자가 system Node 미설치 환경에서도 일단 동작하도록 한다.

import * as fs from 'node:fs'
import * as path from 'node:path'
import { execSync } from 'node:child_process'
import { loggerFor } from '../logging/logger.js'

const log = loggerFor('NodeExecutableLocator')

/** Node 실행 파일 절대 경로 또는 null. 어떤 fallback 단계에서 찾았는지 디버그 로그를 남긴다. */
export function findNodeExecutable(): string | null {
  const found = findFromNvmDefault() ?? findFromNvmVersions() ?? findFromPathFallback()
  if (found) return found

  // 최후 fallback — extension host 가 돌고 있는 Electron Node. better-sqlite3 ABI 가
  // VS Code 버전과 정확히 맞으면 동작, 아니면 어차피 spawn 후 die 한다.
  const electronNode = process.execPath
  if (electronNode && fs.existsSync(electronNode)) {
    log.warn(`Falling back to extension host's process.execPath (${electronNode}) — system Node not found`)
    return electronNode
  }
  log.warn('Node executable not found in any fallback')
  return null
}

function nvmDir(): string {
  return process.env.NVM_DIR ?? path.join(process.env.HOME ?? '', '.nvm')
}

/** 1) ~/.nvm/alias/default 에 적힌 버전. */
function findFromNvmDefault(): string | null {
  const aliasFile = path.join(nvmDir(), 'alias/default')
  if (!fs.existsSync(aliasFile)) return null
  const versionPrefix = fs.readFileSync(aliasFile, 'utf8').trim()

  const versionsDir = path.join(nvmDir(), 'versions/node')
  if (!fs.existsSync(versionsDir)) return null

  const candidates = fs.readdirSync(versionsDir)
    .filter(name => {
      const ver = name.replace(/^v/, '')
      return ver.startsWith(versionPrefix) && fs.existsSync(path.join(versionsDir, name, 'bin/node'))
    })
    .sort()
    .reverse()

  if (candidates.length === 0) return null
  const found = path.join(versionsDir, candidates[0]!, 'bin/node')
  log.debug(`Node found via nvm default alias (${found})`)
  return found
}

/** 2) ~/.nvm/versions/node 스캔 — v24 우선, 그 외 버전 내림차순. */
function findFromNvmVersions(): string | null {
  const versionsDir = path.join(nvmDir(), 'versions/node')
  if (!fs.existsSync(versionsDir)) return null

  const candidates = fs.readdirSync(versionsDir)
    .filter(name => fs.existsSync(path.join(versionsDir, name, 'bin/node')))
    .sort((a, b) => {
      const aMajor = parseInt(a.replace(/^v/, '').split('.')[0]!, 10)
      const bMajor = parseInt(b.replace(/^v/, '').split('.')[0]!, 10)
      const aV24 = aMajor === 24 ? 1 : 0
      const bV24 = bMajor === 24 ? 1 : 0
      if (aV24 !== bV24) return bV24 - aV24
      return b.localeCompare(a)
    })

  if (candidates.length === 0) return null
  const found = path.join(versionsDir, candidates[0]!, 'bin/node')
  log.debug(`Node found via nvm versions scan (${found})`)
  return found
}

/** 3) which/where node — fnm/Homebrew 경로 보강. */
function findFromPathFallback(): string | null {
  const home = process.env.HOME ?? ''
  const extras = [
    path.join(home, '.local/share/fnm/aliases/default/bin'),
    '/usr/local/bin',
    '/opt/homebrew/bin',
  ]
  const isWindows = process.platform === 'win32'
  const sep = isWindows ? ';' : ':'
  const pathEnv = (process.env.PATH ?? '') + sep + extras.join(sep)
  const cmd = isWindows ? 'where node' : 'which node'

  try {
    const out = execSync(cmd, { env: { ...process.env, PATH: pathEnv }, encoding: 'utf8' })
    const first = out.split('\n')[0]?.trim()
    if (first) {
      log.debug(`Node found via PATH fallback (${first})`)
      return first
    }
  } catch (e) {
    log.debug('which/where node failed', e)
  }
  return null
}
